packages/core/src/command/consolidated/consolidated-filter-shared.ts

```diff
@@ -0,0 +1,132 @@
+import { Period, Range } from "@medplum/fhirtypes";
+import { filterTruthy } from "@metriport/shared/common/filter-map";
+import dayjs from "dayjs";
+import customParseFormat from "dayjs/plugin/customParseFormat";
+import isSameOrAfter from "dayjs/plugin/isSameOrAfter";
+import isSameOrBefore from "dayjs/plugin/isSameOrBefore";
+
+dayjs.extend(isSameOrAfter);
+dayjs.extend(isSameOrBefore);
+dayjs.extend(customParseFormat);
+
+export type DateRange = {
+  dateFrom?: string | undefined;
+  dateTo?: string | undefined;
+};
+
+export function arePeriodsWithinRange(
+  periods: (Period | undefined)[] | undefined,
+  range: DateRange
+): boolean | undefined {
+  // we need to check if the periods are valid before we call `.some`, b/c that will always return
+  // boolean and we need to return undefined if there are no valid periods to check - so upstream
+  // functions can check other fields
+  const periodsWithSomeDate = periods?.flatMap(p => (p?.start || p?.end ? p : [])) ?? [];
+  if (periodsWithSomeDate.length < 1) return undefined;
+  return periodsWithSomeDate.some(period => isPeriodWithinRange(period, range));
+}
+export function isPeriodWithinRange(
+  period: Period | undefined,
+  range: DateRange
+): boolean | undefined {
+  const dates = getDatesFromPeriod(period);
+  return areDatesWithinRange(dates, range);
+}
+
+export function areRangesWithinRange(
+  ranges: (Range | undefined)[] | undefined,
+  range: DateRange
+): boolean | undefined {
+  const periodsWithSomeDate = ranges?.flatMap(p => (p?.low || p?.high ? p : [])) ?? [];
+  if (periodsWithSomeDate.length < 1) return undefined;
+  const areRangesWithinRange = periodsWithSomeDate.some(currRange => {
+    const dates = getDateFromRange(currRange);
+    return areDatesWithinRange(dates, range);
+  });
+  return areRangesWithinRange;
+}
+
+/**
+ * Checks if any date in the array is within the date range.
+ * @param dates - An array of dates to check.
+ * @param range - The date range to check against.
+ * @returns - True if any date is within the range, false otherwise.
+ */
+export function areDatesWithinRange(
+  dates: (string | undefined)[],
+  range: DateRange
+): boolean | undefined {
+  // we need to check if the dates are valid before we call `.some`, b/c that will always return boolean
+  // and we need to return undefined if there are no valid dates to check - so upstream functions can
+  // check other fields
+  const parsedDates = dates.flatMap(date => safeDate(date) ?? []);
+  if (parsedDates.length < 1) return undefined;
+  return parsedDates.some(date => isDateWithinDateRange(date, range));
+}
+/**
+ * Checks the date is within the date range.
+ * @param date - The date to check.
+ * @param range - The date range to check against.
+ * @returns - True if the date is within the range, false otherwise.
+ */
+export function isDateWithinDateRange(
```

Comment:
do we need to do something explicit here to make sure this is in utc? 
We have had issues with this sort of thing before (especially in unit tests running local vs on github) and would be good to make everything always utc imo

Reply:
so replace dayjs to always be dayjs.utc

Reply:
All dates are supposed to be UTC, regardless of the environment we run. If we try to convert the input date to UTC, it with either:
- be modified to apply discount the local TZ; or
- force the date as UTC even though it might not be one.

Whatever place that inputs the date should be the one minding TZ, IMO.

Reply:
That's a good point. In theory all dates should be used with TZ info or assumed to be UTC, so for local dev it would be better to have it `dayjs.utc` (I remember doing this in a previous project).

I created the `buildDayjs` so we can build them with the proper initialization (dayjs.utc in this case).

The thing is, we have _a lot_ of places using plain `dayjs(..)`... we'll need to update as we go. Can you add this to the Notion dev guidelines and share with the team, please? I'm happy to help reviewing it.

Reply:
Updated dev guidelines in notion https://www.notion.so/metriport/Development-Best-Practices-bb3574d3c83d46dc9914a9e5c5542ca7?pvs=4#4be8f2115e894912b5e6896e72c4c6c5

Reply:
Awesome, I reviewed it and it looks good!